Respect exclude option when handling url

Is it possible to respect exclude option when handling url?

I have a specific use case, in our CMS an url that has '++resource++' inside should be considered external image. This is my current configuration:

```javascript
     {
        // Handle SCSS files
        test: /\.scss$/,
        exclude: /\+\+resource\+\+/,
        use: ExtractTextPlugin.extract({
          fallback: 'style-loader',
          use: [
            'css-loader',
            'postcss-loader',
            'sass-loader'
          ]
        }),
      },
```

But the url handling is not respecting this configuration, and throw this traceback:
```javascript
    ERROR in ./node_modules/css-loader!./node_modules/postcss-loader/lib!./node_modules/sass-loader/lib/loader.js!./app/cover.scss
    Module not found: Error: Can't resolve './++resource++plone.app.jquerytools.pb_close.png' in '/home/rodfersou/.projects/collective.cover/webpack/app'
     @ ./node_modules/css-loader!./node_modules/postcss-loader/lib!./node_modules/sass-loader/lib/loader.js!./app/cover.scss 7:6933-6992
```

--- a/README.md
+++ b/README.md
@@ -114,23 +114,41 @@ module.exports = {
 
 ## Options
 
-|                    Name                     |        Type         |     Default     | Description                                 |
-| :-----------------------------------------: | :-----------------: | :-------------: | :------------------------------------------ |
-|              **[`url`](#url)**              |     `{Boolean}`     |     `true`      | Enable/Disable `url()` handling             |
-|           **[`import`](#import)**           |     `{Boolean}`     |     `true`      | Enable/Disable @import handling             |
-|          **[`modules`](#modules)**          | `{Boolean\|String}` |     `false`     | Enable/Disable CSS Modules and setup mode   |
-|   **[`localIdentName`](#localidentname)**   |     `{String}`      | `[hash:base64]` | Configure the generated ident               |
-|        **[`sourceMap`](#sourcemap)**        |     `{Boolean}`     |     `false`     | Enable/Disable Sourcemaps                   |
-|        **[`camelCase`](#camelcase)**        | `{Boolean\|String}` |     `false`     | Export Classnames in CamelCase              |
-|    **[`importLoaders`](#importloaders)**    |     `{Number}`      |       `0`       | Number of loaders applied before CSS loader |
-| **[`exportOnlyLocals`](#exportonlylocals)** |     `{Boolean}`     |     `false`     | Export only locals                          |
+|                    Name                     |         Type          |     Default     | Description                                 |
+| :-----------------------------------------: | :-------------------: | :-------------: | :------------------------------------------ |
+|              **[`url`](#url)**              | `{Boolean\|Function}` |     `true`      | Enable/Disable `url()` handling             |
+|           **[`import`](#import)**           |      `{Boolean}`      |     `true`      | Enable/Disable @import handling             |
+|          **[`modules`](#modules)**          |  `{Boolean\|String}`  |     `false`     | Enable/Disable CSS Modules and setup mode   |
+|   **[`localIdentName`](#localidentname)**   |      `{String}`       | `[hash:base64]` | Configure the generated ident               |
+|        **[`sourceMap`](#sourcemap)**        |      `{Boolean}`      |     `false`     | Enable/Disable Sourcemaps                   |
+|        **[`camelCase`](#camelcase)**        |  `{Boolean\|String}`  |     `false`     | Export Classnames in CamelCase              |
+|    **[`importLoaders`](#importloaders)**    |      `{Number}`       |       `0`       | Number of loaders applied before CSS loader |
+| **[`exportOnlyLocals`](#exportonlylocals)** |      `{Boolean}`      |     `false`     | Export only locals                          |
 
 ### `url`
 
-Type: `Boolean`
+Type: `Boolean|Function`
 Default: `true`
 
-Enable/disable `url()` resolving. Absolute `urls` are not resolving by default.
+Control `url()` resolving. Absolute `urls` are not resolving by default.
+
+Examples resolutions:
+
+```
+url(image.png) => require('./image.png')
+url(./image.png) => require('./image.png')
+```
+
+To import assets from a `node_modules` path (include `resolve.modules`) and for `alias`, prefix it with a `~`:
+
+```
+url(~module/image.png) => require('module/image.png')
+url(~aliasDirectory/image.png) => require('otherDirectory/image.png')
+```
+
+#### `Boolean`
+
+Enable/disable `url()` resolving.
 
 **webpack.config.js**
 
@@ -150,18 +168,27 @@ module.exports = {
 };
 ```
 
-Examples resolutions:
+#### `Function`
 
-```
-url(image.png) => require('./image.png')
-url(./image.png) => require('./image.png')
-```
+Allow to filter `url()`. All filtered `url()` will not be resolved.
 
-To import assets from a `node_modules` path (include `resolve.modules`) and for `alias`, prefix it with a `~`:
-
-```
-url(~module/image.png) => require('module/image.png')
-url(~aliasDirectory/image.png) => require('otherDirectory/image.png')
+```js
+module.exports = {
+  module: {
+    rules: [
+      {
+        test: /\.css$/,
+        loader: 'css-loader',
+        options: {
+          url: (url, resourcePath) => {
+            // `url()` with `img.png` stay untouched
+            return url.includes('img.png');
+          },
+        },
+      },
+    ],
+  },
+};
 ```
 
 ### `import`

--- a/src/index.js
+++ b/src/index.js
@@ -26,6 +26,7 @@ import {
   getImportPrefix,
   placholderRegExps,
   dashesCamelCase,
+  getFilter,
 } from './utils';
 import Warning from './Warning';
 import CssSyntaxError from './CssSyntaxError';
@@ -66,9 +67,6 @@ export default function loader(content, map, meta) {
     }
   }
 
-  const resolveImport = options.import !== false;
-  const resolveUrl = options.url !== false;
-
   const plugins = [];
 
   if (options.modules) {
@@ -100,14 +98,16 @@ export default function loader(content, map, meta) {
     );
   }
 
-  if (resolveImport) {
+  if (options.import !== false) {
     plugins.push(importParser());
   }
 
-  if (resolveUrl) {
+  if (options.url !== false) {
     plugins.push(
       urlParser({
-        filter: (value) => isUrlRequest(value),
+        filter: getFilter(options.url, this.resourcePath, (value) =>
+          isUrlRequest(value)
+        ),
       })
     );
   }

--- a/src/options.json
+++ b/src/options.json
@@ -2,7 +2,14 @@
   "additionalProperties": false,
   "properties": {
     "url": {
-      "type": "boolean"
+      "anyOf": [
+        {
+          "type": "boolean"
+        },
+        {
+          "instanceof": "Function"
+        }
+      ]
     },
     "import": {
       "type": "boolean"

--- a/src/utils.js
+++ b/src/utils.js
@@ -61,4 +61,24 @@ function getLocalIdent(loaderContext, localIdentName, localName, options) {
     .replace(/^((-?[0-9])|--)/, '_$1');
 }
 
-export { getImportPrefix, getLocalIdent, placholderRegExps, dashesCamelCase };
+function getFilter(filter, resourcePath, defaultFilter = null) {
+  return (content) => {
+    if (defaultFilter && !defaultFilter(content)) {
+      return false;
+    }
+
+    if (typeof filter === 'function') {
+      return !filter(content, resourcePath);
+    }
+
+    return true;
+  };
+}
+
+export {
+  getImportPrefix,
+  getLocalIdent,
+  placholderRegExps,
+  dashesCamelCase,
+  getFilter,
+};

--- a/test/errors.test.js
+++ b/test/errors.test.js
@@ -22,6 +22,7 @@ it('validation', () => {
 
   expect(() => validate({ url: true })).not.toThrow();
   expect(() => validate({ url: false })).not.toThrow();
+  expect(() => validate({ url: () => {} })).not.toThrow();
   expect(() => validate({ url: 'true' })).toThrowErrorMatchingSnapshot();
 
   expect(() => validate({ import: true })).not.toThrow();

--- a/test/url-option.test.js
+++ b/test/url-option.test.js
@@ -54,4 +54,31 @@ describe('url option', () => {
       );
     });
   });
+
+  it('Function', async () => {
+    const config = {
+      loader: {
+        options: {
+          url: (url, resourcePath) => {
+            expect(typeof resourcePath === 'string').toBe(true);
+
+            return url.includes('img.png');
+          },
+        },
+      },
+    };
+    const testId = './url/url.css';
+    const stats = await webpack(testId, config);
+    const { modules } = stats.toJson();
+    const module = modules.find((m) => m.id === testId);
+
+    expect(module.source).toMatchSnapshot('module');
+    expect(evaluated(module.source, modules)).toMatchSnapshot(
+      'module (evaluated)'
+    );
+    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
+      'warnings'
+    );
+    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
+  });
 });

